File: app/src/main/java/net/ginteam/carmen/presenter/company/CompaniesPresenter.java
package net.ginteam.carmen.presenter.company;

import android.util.Log;

import net.ginteam.carmen.R;
import net.ginteam.carmen.contract.company.CompaniesContract;
import net.ginteam.carmen.manager.FiltersViewStateManager;
import net.ginteam.carmen.manager.PreferencesManager;
import net.ginteam.carmen.manager.SortViewStateManager;
import net.ginteam.carmen.model.Pagination;
import net.ginteam.carmen.model.city.CityModel;
import net.ginteam.carmen.model.company.CompanyModel;
import net.ginteam.carmen.provider.ModelCallback;
import net.ginteam.carmen.provider.ModelCallbackWithMeta;
import net.ginteam.carmen.provider.auth.AuthProvider;
import net.ginteam.carmen.provider.company.CompaniesProvider;
import net.ginteam.carmen.provider.company.FavoritesProvider;

import java.util.List;

/**
 * Created by Eugene on 12/27/16.
 */

public class CompaniesPresenter implements CompaniesContract.Presenter {

    private CompaniesContract.View mView;
    private boolean mIsFirstLoading;

    @Override
    public String getUserCityName() {
        return PreferencesManager
                .getInstance()
                .getCity().getName();
    }

    @Override
    public void selectCompany(CompanyModel company) {

    }

    @Override
    public void fetchRecentlyWatchedCompanies() {
        mView.showLoading(true);

        CompaniesProvider
                .getInstance()
                .fetchRecentlyWatched(new ModelCallback<List<CompanyModel>>() {
                    @Override
                    public void onSuccess(List<CompanyModel> resultModel) {
                        mView.showLoading(false);
                        mView.showCompanies(resultModel, null);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void fetchPopularCompanies() {
        mView.showLoading(true);

        CityModel userCity = PreferencesManager.getInstance().getCity();
        CompaniesProvider
                .getInstance()
                .fetchPopular(userCity.getId(), new ModelCallback<List<CompanyModel>>() {
                    @Override
                    public void onSuccess(List<CompanyModel> resultModel) {
                        mView.showLoading(false);
                        mView.showCompanies(resultModel, null);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void fetchFavoriteCompanies() {
        mView.showLoading(true);

        FavoritesProvider
                .getInstance()
                .fetchFavorite(new ModelCallback<List<CompanyModel>>() {
                    @Override
                    public void onSuccess(List<CompanyModel> resultModel) {
                        mView.showLoading(false);
                        mView.showCompanies(resultModel, null);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void fetchCompaniesForCategory(int categoryId, String filter, String sortField,
                                          String sortType, int page) {
        mIsFirstLoading = (page == 1);
        if (mIsFirstLoading) {
            mView.showLoading(true);
        }

        CompaniesProvider
                .getInstance()
                .fetchForCategory(categoryId, filter, sortField, sortType, page, new ModelCallbackWithMeta<List<CompanyModel>>() {
                    @Override
                    public void onSuccess(List<CompanyModel> resultModel, Pagination pagination) {
                        mView.showLoading(false);
                        if (mIsFirstLoading) {
                            mView.showCompanies(resultModel, pagination);
                            mIsFirstLoading = false;
                            return;
                        }
                        mView.showMoreCompanies(resultModel);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void addToFavoriteClick(CompanyModel companyModel) {
        if (AuthProvider.getInstance().getCurrentCachedUser() != null) {
            if (!companyModel.isFavorite()) {
                companyModel.setFavorite(true);
                mView.addToFavorites();
                addToFavorites(companyModel);
                return;
            }
            companyModel.setFavorite(false);
            mView.removeFromFavorites(companyModel);
            removeFromFavorites(companyModel);
            return;
        }
        mView.showError(mView.getContext().getResources().getString(R.string.access_denied_message));
    }

    @Override
    public void addToFavorites(CompanyModel company) {
        FavoritesProvider
                .getInstance()
                .addToFavorites(company, new ModelCallback<String>() {
                    @Override
                    public void onSuccess(String resultModel) {
                        Log.d("COMPANIES_PRESENTER", "Add to favorite success");
                    }

                    @Override
                    public void onFailure(String message) {
                        Log.e("COMPANIES_PRESENTER", "Add to favorite error: " + message);
                    }
                });
    }

    @Override
    public void removeFromFavorites(final CompanyModel company) {
        FavoritesProvider
                .getInstance()
                .removeFromFavorites(company, new ModelCallback<String>() {
                    @Override
                    public void onSuccess(String resultModel) {
                        Log.d("COMPANIES_PRESENTER", "Remove from favorite success");
                    }

                    @Override
                    public void onFailure(String message) {
                        Log.e("COMPANIES_PRESENTER", "Remove from favorite error: " + message);
                    }
                });
    }

    @Override
    public void attachView(CompaniesContract.View view) {
        mView = view;
        FiltersViewStateManager.getInstance().resetFiltersState();
        SortViewStateManager.getInstance().resetFiltersState();
    }

    @Override
    public void detachView() {
        mView = null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/presenter/company/CompaniesPresenter.kt
package net.ginteam.carmen.kotlin.presenter.company

import net.ginteam.carmen.kotlin.api.response.MetaSubscriber
import net.ginteam.carmen.kotlin.contract.CompaniesContract
import net.ginteam.carmen.kotlin.contract.SortContract
import net.ginteam.carmen.kotlin.manager.PreferencesManager
import net.ginteam.carmen.kotlin.manager.SharedPreferencesManager
import net.ginteam.carmen.kotlin.model.CompanyModel
import net.ginteam.carmen.kotlin.model.PaginationModel
import net.ginteam.carmen.kotlin.view.fragment.sort.SortOptionDialogFragmentViewState

/**
 * Created by eugene_shcherbinock on 2/16/17.
 */
class CompaniesPresenter : BaseCompaniesPresenter <CompaniesContract.View>(), CompaniesContract.Presenter {

    private var isFirsLoading: Boolean = true

    private val mPreferences: PreferencesManager = SharedPreferencesManager

    override fun fetchCompanies(categoryId: Int, filterQuery: String, sortField: String, sortType: String, pageNumber: Int) {
        isFirsLoading = pageNumber == 1
        if (isFirsLoading) {
            mView?.showLoading(true)
        }

        val filter = filterQuery.plus("cityId:${mPreferences.userCityModel!!.id}")
        mCompaniesProvider
                .fetchCompanies(categoryId, filter, sortField, sortType, pageNumber)
                .subscribe(object : MetaSubscriber <MutableList <CompanyModel>>() {
                    override fun success(model: MutableList<CompanyModel>, pagination: PaginationModel) {
                        mView?.showLoading(false)
                        if (isFirsLoading) {
                            mView?.showCompanies(model, pagination)
                            return
                        }
                        mView?.showCompanies(model)
                    }

                    override fun error(message: String, isNetworkError: Boolean) {
                        mView?.showError(message, isNetworkError)
                    }
                })
    }

    override fun attachView(view: CompaniesContract.View) {
        super.attachView(view)

        // reset view states when fragment reattached
        val sortViewState: SortContract.SortViewState = SortOptionDialogFragmentViewState
        sortViewState.resetViewState()
    }

}